### PR TITLE
Environment for architecture and training parameters selection for Deep CNN training

### PR DESCRIPTION
**Pull request by iaroslav-ai**
These commits contain an environment as discussed in #94. In this environment agent can select an architecture of deep CNN classifier and training parameters such that it results in high accuracy. 

It was originally discussed to start with more “classical” methods like SVM etc, but I think it is better to start directly with Deep Learning as parameter selection for the “classical” methods is usually relatively straightforward.

One step in this environment is a training of a deep network for 10 epochs, where architecture and training parameters are selected by an agent.  One episode in this
environment has fixed size of 10 steps. Reward is proportional to the best validation accuracy achieved at current step.

Training parameters that agent can adjust are learning rate, momentum, batch size, l1 and l2 penalty coefficients. Agent can select up to 5 layers of CNN and up to 2 layers of fully connected layers.  As an observation, agent receives # of instances in a dataset and a validation accuracy at every step.

At the beginning of every episode, a dataset to train on is chosen at random. Datasets used are MNIST, CIFAR10, CIFAR100. Number of instances in datasets are chosen at random in range from around 100% to 5% such that adjustment of l1, l2 penalty coefficients makes more difference.

Some of the parameters of the dataset are not provided to the agent in order to make
agent figure it out through experimentation during an episode.

This environment requires Keras with Theano or TensorFlow to run. When run on laptop gpu (GTX960M) one step takes on average 2 min.

### CODE REVIEW

**Comment by jietang** [comment]
Thanks @iaroslav-ai, merged.

One potential followup: I think it's likely that current RL algorithms won't be able to handle an env that takes 2 minutes per call to step. Many of them involve taking multiple rollouts (i.e. until done=True) and then computing a single gradient update. Maybe there's a way to deliver more feedback as training progresses?

**Comment by iaroslav-ai** [comment]
Right, RL methods that are popular right now require experimentation with environment to be cheap, and therefore I would also expect that this kind of environment will be tough for these methods. 

There are a few ways I could imagine to address this:

Mix this environment with convergence control, such that agent could both set the architecture of the deep net at the beginning of the training, and control / observe convergence. This could potentially increase number of reward evaluations by an order of magnitude, for around the same time cost. 

Also I could imagine an environment (or set of envs) where you start with smaller amounts of data and as agent becomes better, switch to larger data. I could imagine that this would help speed up agent learning at least in the early training stages, due to training being faster on small data.

Also model based RL approaches can be used; Here is one nice tutorial on these methods that I found http://videolectures.net/nips09_littman_mbrl/ if you are not familiar with those. According to some figures there (55:06), model based RL could be more than order of magnitude more data efficient than model free one. However, model based RL would require to build a model of what training results are given architecture / training parameters; such model would require to have stochastic outputs, due to the stochastic nature of training of NN. 

High cost of experimentation with environment is actually one of the reasons I think this environment is interesting, as it requires data efficient RL methods.

I think the next thing that I will do is I will run model free RL methods on this environment, and see how they behave. Would be also interesting to compare model free vs model based RL methods.

**Comment by jietang** [comment]
Yes I agree with your reasoning. One thing I would add is that while high cost of experimentation is interesting, extremely high cost makes it difficult to experiment. For example CartPole takes ~100 rollouts to solve with ~100 timesteps per rollout; if each step took 2 minutes, that's about 2 weeks of compute time for a single training run, and cartpole is on the low end in terms of timesteps and # of rollouts to solve.

**Comment by iaroslav-ai** [comment]
Right, this is a good point.
One thing I could imagine is to train a neural network, which given architecture, training parameters, and any other parameters which make sense estimates what the outcome of the training would be. To build such network, one could run training many times with different (e.g. randomly sampled) architecture / training params etc, thus collecting a dataset, which is then used to train a neural net. 
Then such network can be used internally instead of actual neural network training, provided that it models the validation accuracy of training outcome sufficiently accurately. The good thing is that such neural net needs to be trained only once. 
I could imagine that one can train first RL agent using such computationally efficient version of environment, and then test on real thing, to see if it generalizes. 
Provided that behavior of training can be modeled accurately enough with nn of reasonable size (or any other class of ML models, whatever works best), I would expect that the step cost would dramatically decrease.

**Comment by jietang** [comment]
Yes, that's an interesting idea; you could view it as a model-based RL algorithm, where you want to make better use of expensive real-world experience by fitting a (NN) model and training your RL agent on that.

**Comment by iaroslav-ai** [comment]
Right. Thanks for your feedback! :)
